### PR TITLE
Custom snapshots

### PR DESCRIPTION
**Pull request by sigus**
Primitive implementation of #1990. No more autosaves and backup per player name, just one autosave and one backup. Instead, player can add custom snapshots.

### CODE REVIEW

**Review by bobbens — CHANGES_REQUESTED** [review]
Thanks for the PR!
 I feel that the implementation is a bit too dangerous, as it seems like you can easily overwrite existing progress. In particular, this is because there seems to be a single global `backup` and `autosave` file. From my understanding, you could do something like:
1. Load 'foo'
2. Play and have it update 'autosave'
3. Load 'bar' (not 'autosave')
4. Game saves and now your progress in 'foo' that was saved in 'autosave' got overwritten

I also think snapshot management would be best to do directly from the load menu instead of the small in-game menu. As you are likely to create a snapshot backup before loading another snapshot, and it would be more natural that way.

I think it might be best to have the autosaves, backups and snapshots tied to the player save (at least in the menu, they could be physically all separate). So you can load a game directly (like now), or have an option to `load snapshot` which would let you choose what snapshot to load (including autosave (default and gets overwritten on save), backup or any arbitrary snapshots). It should also give a warning that autosave will get overwritten when loading another snapshot. 

I think a good example of how it probably should be done is the Endless Sky implementation, which is more or less how I explained above. I'm attaching a screenshot for potential reference.
es_save

File: src/load.c [hunk 1]
```diff
@@ -230,6 +213,10 @@ int load_refresh (void)
          ns = &array_grow( &load_saves );
       snprintf( buf, sizeof(buf), "saves/%s", files[i].name );
       ok = load_load( ns, buf );
+      len = MIN( strlen( files[i].name ) - 3, sizeof(buf) - 1 );
+      strncpy( buf, files[i].name, len );
```
[the remark this answers is not in the document]
Comment (sigus) [reply src/load.c:217]: This is a clumsy attempt to cut off the file extension.

[the remark this answers is not in the document]
Comment (bobbens) [reply src/load.c:217]: I see, wouldn't something like the following suffice?
```
ns->save_name = strdup( files[i].name );
ns->save_name[ strlen( ns->save_name )-3 ] = '\0';
```

**Comment by sigus** [comment]
> I think it might be best to have the autosaves, backups and snapshots tied to the player save (at least in the menu, they could be physically all separate). 

> I think a good example of how it probably should be done is the Endless Sky implementation, which is more or less how I explained above.

In Endless Sky you can create new pilots and their saves will be shown separately. Do you want something like this in Naev?

**Comment by bobbens** [comment]
> In Endless Sky you can create new pilots and their saves will be shown separately. Do you want something like this in Naev?

Creating new pilots can be done with the main menu, I don't think it has to be done through the loading interface.

For the interface, I think having a list with one line per pilot like now (defaulting to that pilots autosave) would be good. All that would have to be added is some "snapshots" button that would let you manipulate snapshots for a pilot, by opening a new window with a list of their snapshots where you can delete, add, or load snapshots.

From a player perspective, the only change would be that in the load menu there is a snapshot button that lets you manipulate snapshots, other than that, it would behave the same as now.

**Comment by sigus** [comment]
> I also think snapshot management would be best to do directly from the load menu instead of the small in-game menu.

However, it makes sense to be able to open the load menu from the small menu. But allow saving custom snapshots (and autosaves and backups) only when player can save (saving not disabled via Lua API or in some other way), as discussed in Discord earlier. Presumably, you should first start or load the game in order to save it.

> I think it might be best to have the autosaves, backups and snapshots tied to the player save (at least in the menu, they could be physically all separate).

Currently, the game warns about overwrite when you create new pilot with existing name. With custom snapshots the player's name becomes a kind of "directory" for saved games which in fact can be completely different games.

> For the interface, I think having a list with one line per pilot like now (defaulting to that pilots autosave) would be good.

> From a player perspective, the only change would be that in the load menu there is a snapshot button that lets you manipulate snapshots, other than that, it would behave the same as now.

So where should auto-backups be shown, where are they now or they should be hidden behind "snapshots" button and only player names (which default to autosave) should be shown in the load menu?

File: src/load.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (sigus) [reply src/load.c:217]: Yes, that's suffice. Only 3 chars more memory will be retained until freed.

Is there any way to mention who suggested the changes in the commit besides the commit message? It seems that there is some kind of functionality in git and/or GitHub for this.

**Comment by bobbens** [comment]
> > I also think snapshot management would be best to do directly from the load menu instead of the small in-game menu.
> 
> However, it makes sense to be able to open the load menu from the small menu. But allow saving custom snapshots (and autosaves and backups) only when player can save (saving not disabled via Lua API or in some other way), as discussed in Discord earlier. Also, presumably, you should first start or load the game in order to save it.

Yeah, that can work too.

> > I think it might be best to have the autosaves, backups and snapshots tied to the player save (at least in the menu, they could be physically all separate).
> 
> Currently, the game warns about overwrite when you create new pilot with existing name. With custom snapshots the player's name becomes a kind of "directory" for saved games which in fact can be completely different games.

Yeah, implementing it as a directory per player does seem like it makes most sense, but we'll need some backward compatibility to load old saves.

> > For the interface, I think having a list with one line per pilot like now (defaulting to that pilots autosave) would be good.
> 
> > From a player perspective, the only change would be that in the load menu there is a snapshot button that lets you manipulate snapshots, other than that, it would behave the same as now.
> 
> So where should auto-backups be shown, where are they now or they should be hidden behind "snapshots" button and only player names (which default to autosave) should be shown in the load menu?

Yes. I think hiding backups behind snapshots would be simplest and most coherent.

**Comment by sigus** [comment]
> we'll need some backward compatibility to load old saves

What kind of backward compatibility fits best? I'm thinking about some variants, but none of them seems good:

1. Just iterate over old saves and add them to list, but new autosaves will be written in corresponding subdirectories and old will persist and make confusion
2. Silently move old saves to subdirectories, but unclear overwrite or not if corresponding autosave in subdirectory already exists
3. Prompt user if they want to move old saves and maybe overwrite existing saves in subdirectories? But what to do if they decline?
4. Don't use subdirectories and use some kind of filename prefixes like Endless Sky. But still have to rename backups? It also has its own subtleties.
5. ???

Presumably there shouldn't be anything in subdirectories initially, but...

**Comment by bobbens** [comment]
What about doing 2, but first backing up the old saves to `saves-pre-0.10.0`, and informing the user that saves have been upgraded, and the old saves are backed up if they wish to do anything? That might be the best option.

File: src/load.c [hunk 2]
```diff
@@ -277,6 +309,83 @@ static int load_enumerateCallback( void* data, const char* origdir, const char*
...
+         free( path );
+         return PHYSFS_ENUM_OK;
+      }
+      if (!PHYSFS_exists( "saves-pre-0.10.0" ))
+         PHYSFS_mkdir( "saves-pre-0.10.0" );
+      asprintf( &backup_path, "saves-pre-0.10.0/%s", fname );
+      if (!ndata_copyIfExists( path, backup_path ))
+         old_saves_detected = 1;
+      free( backup_path );
+      if (name_len >= 4 && !strcmp( &fname[name_len-3], ".ns" )) {
```
Comment (bobbens) [inline src/load.c:342]: You can probably consolidate the block here and the next one for `.ns.backup` into an external function to reduce code duplication.

**Comment by bobbens** [comment]
This seems mainly ready, is there anything still missing?

**Comment by sigus** [comment]
This is what I see now:
1. Warning about old saves backed up on black screen before main menu appears. Also the warning text could be better.
2. Some bugs including SIGSEGVs.
3. Can load a snapshot form the small menu only for the current player.
4. Some "desynchronization" between "Load Pilot" and "Load Snapshot" menus (and filesystem: empty dirs not deleted etc).
5. There is no distinction between autosave and manual explicit rewrite of autosave when save_loaded != 0. Not sure if it should be.
6. Blinking with that new fading stuff? Not sure.

**Comment by bobbens** [comment]
Looks good, thanks!

**Pull request merged**